### ISSUE TITLE
System.Console.ReadLine infinite loop under WASM

### ISSUE THREAD

**Issue by masonwheeler** [mono/mono#9058]
[I reported this over in Blazor](https://github.com/aspnet/Blazor/issues/945) and was informed that the API comes from Mono and needed to be reported here.

When using a Blazor project, which runs the Mono WASM interpreter in a web browser, a call to `System.Console.ReadLine` should give the user a single prompt.  Instead, it gives an infinite loop of prompts until the user tells the browser to prevent the page from opening more prompts, and then yields the value entered into the first prompt as the return value, discarding the inputs from subsequent prompts.

This behavior can be verified by running the following page under Blazor:

```
@page "/"

<h1>Hello, @(_name)!</h1>

Welcome to your new app.<br />

<button onclick=@ReadInput>Enter Name</button>

@functions {
	private string _name = "world";

	private void ReadInput() {
		_name = System.Console.ReadLine();
		StateHasChanged();
	}
}
```

**Comment by kumpera** [comment]
Hi,

Thanks for reporting this issue. It took me a bit to realize that you’re trying to use ReadLine.
That’s kind of unexpected as it’s not the most intuitive way of asking for user input.

As a workaround, while we take a look at fixing this, you can use the JS prompt function directly.

**Comment by masonwheeler** [comment]
I ran into it while playing around with [Runny](https://github.com/Suchiman/Runny), which runs the Roslyn compiler inside of the browser with Blazor.  Because the compiler doesn't know about the Blazor environment, Console input really is the only way available in that specific case.

**Comment by masonwheeler** [comment]
The problem appears to be in `TTY.streamOps.read` in mono.js, where it tries to keep reading chars from input until it's filled a 1024-character buffer, with no provision to bail out early if it hits an end-of-line.

**Comment by Stuart88** [comment]
+1 for this.

I'm trying to get to grips with Blazor, attempting to make one of my old C# console apps work in the browser.

It has a lot of Console.ReadLine calls for getting user input, so I've run into the same problem. Everything else works well apart from this.

I'm going to try to to find a workaround, but yes it would definitely be nice to have ReadLine() not looping forever!

**Comment by mfe-** [comment]
@masonwheeler did you find a workaround for this?

**Comment by Stuart88** [comment]
> @masonwheeler, @Stuart88 did you find a workaround for this?

Nope, sorry! I was just playing around, I probably just gave up.

**Comment by masonwheeler** [comment]
No. If there were any obvious workaround, I wouldn't have needed to report a bug.  I take it, then, that this broken implementation has never received any attention?

**Comment by akoeplinger** [comment]
No and `Console.ReadLine()` in .NET 5 throws PlatformNotSupportedException now. We'd need to open a new issue in https://github.com/dotnet/runtime if we want to properly hook up the JS prompt.

**Comment by mfe-** [comment]
I might have found a workaround for this issue. I was playing around with the goal to port existing Console Applications to Blazor [here](https://github.com/mfe-/Blazor.Console/).

I set `System.Console.SetIn();` to my own custom `TextReader`. It redirects `System.Console.ReadLine()` calls to my [method](https://github.com/mfe-/Blazor.Console/blob/e88be301d963d36830a525ad9d193391b1cedf64/Console.Extensions.Blazor/BlazorConsole.cs#L105) which invokes the javascript call `prompt`. I'm not finished with experimenting, but from the first look, it seems to be working.  Didn't trie out with .NET 5.
